src/backend/src/lib.rs

```diff
@@ -28,26 +27,34 @@ use shared::types::custom_token::{CustomToken, CustomTokenId};
 use shared::types::token::{UserToken, UserTokenId};
 use shared::types::transaction::SignRequest;
 use shared::types::user_profile::{
-    AddUserCredentialRequest, GetUsersRequest, GetUsersResponse, OisyUser, UserCredential,
+    AddUserCredentialRequest, GetUsersRequest, GetUsersResponse, OisyUser, StoredUserProfile,
     UserProfile,
 };
 use shared::types::{Arg, InitArg, SupportedCredential};
 use std::borrow::Cow;
 use std::cell::RefCell;
 use std::str::FromStr;
+use user_profile::get_or_create;
 
 mod assertions;
 mod guards;
 mod token;
+mod user_profile;
 
 type VMem = VirtualMemory<DefaultMemoryImpl>;
 type ConfigCell = StableCell<Option<Candid<Config>>, VMem>;
 type UserTokenMap = StableBTreeMap<StoredPrincipal, Candid<Vec<UserToken>>, VMem>;
 type CustomTokenMap = StableBTreeMap<StoredPrincipal, Candid<Vec<CustomToken>>, VMem>;
+/// Map of (`updated_timestamp`, `user_principal`) to `UserProfile`
+type UserProfileMap = StableBTreeMap<(u64, StoredPrincipal), Candid<StoredUserProfile>, VMem>;
```

Comment:
Good point. I added the `Timestamp` type but I kept the comment because it specifies which timestamp is used as key. I believe that's important. And having two types `UpdatedTimestamp` and `CreatedTimestamp` is a bit too much, wdyt?

Reply:
Comment is not really useful to me but, for sure I don't mind if it's there 🙂